web/src/scheduled_messages.js

```diff
@@ -8,28 +8,62 @@ import * as compose_ui from "./compose_ui";
 import {$t} from "./i18n";
 import * as narrow from "./narrow";
 import * as notifications from "./notifications";
-import * as overlays from "./overlays";
+import {page_params} from "./page_params";
 import * as people from "./people";
 import * as popover_menus from "./popover_menus";
+import * as stream_data from "./stream_data";
 
-// This is only updated when user opens the scheduled messages overlay.
 export let scheduled_messages_data = [];
 
-export function override_scheduled_messages_data(data) {
-    scheduled_messages_data = data;
+function sort_scheduled_messages_data() {
+    scheduled_messages_data.sort(
+        (msg1, msg2) => msg1.scheduled_delivery_timestamp - msg2.scheduled_delivery_timestamp,
+    );
 }
 
-export function edit_scheduled_message(scheduled_msg_id) {
-    const scheduled_msg = scheduled_messages_data.find(
-        (msg) => msg.scheduled_message_id === scheduled_msg_id,
+export function add_scheduled_messages(scheduled_messages) {
+    scheduled_messages_data.push(...scheduled_messages);
+    sort_scheduled_messages_data();
+}
+
+export function remove_scheduled_message(scheduled_message_id) {
+    const msg_index = scheduled_messages_data.findIndex(
+        (msg) => msg.scheduled_message_id === scheduled_message_id,
     );
+    if (msg_index !== undefined) {
+        scheduled_messages_data.splice(msg_index, 1);
+    }
+}
 
-    let compose_args;
+export function update_scheduled_message(scheduled_message) {
+    const msg_index = scheduled_messages_data.findIndex(
+        (msg) => msg.scheduled_message_id === scheduled_message.scheduled_message_id,
+    );
+
+    if (msg_index === undefined) {
+        return;
+    }
+
+    let needs_to_sorted = false;
+    if (
+        scheduled_messages_data[msg_index].scheduled_delivery_timestamp !==
+        scheduled_message.scheduled_delivery_timestamp
+    ) {
+        needs_to_sorted = true;
+    }
+
+    scheduled_messages_data[msg_index] = scheduled_message;
+    if (needs_to_sorted) {
```

Comment:
`needs_resorting` would be a nicer variable name, but also probably we should just sort unconditionally; it's a really low value optimization.